File: renderer.py
# renderer.py - rendering grid and motifs + animation
import matplotlib.pyplot as plt
from matplotlib import animation
from motifs import loop_around_dot, cross_lines, arc_motif, diagonal_cross, spiral, line_between_dots

def render_grid_with_loops(dot_grid, scale=1.0, style='loop', show=True, figsize=(6,6)):
    fig, ax = plt.subplots(figsize=figsize)
    ax.set_aspect('equal')
    for i, row in enumerate(dot_grid):
        for j, (x, y) in enumerate(row):
            x_s, y_s = x * scale, y * scale
            ax.plot(x_s, y_s, 'ko', markersize=3)
            if style == 'loop':
                xs, ys = loop_around_dot((x_s, y_s), radius=0.35*scale)
                ax.plot(xs, ys)
            elif style == 'cross':
                segs = cross_lines((x_s, y_s), length=0.35*scale)
                for xs, ys in segs:
                    ax.plot(xs, ys)
            elif style == 'arc':
                xs, ys = arc_motif((x_s, y_s), radius=0.4*scale)
                ax.plot(xs, ys)
            elif style == 'diagonal':
                segs = diagonal_cross((x_s, y_s), length=0.35*scale)
                for xs, ys in segs:
                    ax.plot(xs, ys)
            elif style == 'spiral':
                xs, ys = spiral((x_s, y_s), radius=0.4*scale)
                ax.plot(xs, ys)
            elif style == 'tiling':
                # connect right and bottom neighbors if exist
                if j < len(row) - 1:
                    p2 = (row[j+1][0]*scale, row[j+1][1]*scale)
                    xs, ys = line_between_dots((x_s,y_s), p2)
                    ax.plot(xs, ys)
                if i < len(dot_grid) - 1:
                    p2 = (dot_grid[i+1][j][0]*scale, dot_grid[i+1][j][1]*scale)
                    xs, ys = line_between_dots((x_s,y_s), p2)
                    ax.plot(xs, ys)
    ax.axis('off')
    if show:
        plt.show()
    return fig, ax

def animate_kolam(dot_grid, style='loop', scale=1.0, interval=300, figsize=(6,6)):
    fig, ax = plt.subplots(figsize=figsize)
    ax.set_aspect('equal')
    ax.axis('off')
    max_frames = len(dot_grid) * max(len(r) for r in dot_grid)
    artists = []

    def draw_frame(frame):
        ax.clear()
        ax.set_aspect('equal')
        ax.axis('off')
        # draw dots always
        for i, row in enumerate(dot_grid):
            for j, (x, y) in enumerate(row):
                ax.plot(x*scale, y*scale, 'ko', markersize=3)
        # draw up to this frame
        frame_i = frame // max(len(r) for r in dot_grid)
        frame_j = frame % max(len(r) for r in dot_grid)
        count = frame + 1
        idx = 0
        for i, row in enumerate(dot_grid):
            for j, (x, y) in enumerate(row):
                if idx >= count:
                    break
                x_s, y_s = x*scale, y*scale
                if style == 'loop':
                    xs, ys = loop_around_dot((x_s, y_s), radius=0.35*scale)
                    ax.plot(xs, ys)
                elif style == 'spiral':
                    xs, ys = spiral((x_s, y_s), radius=0.4*scale)
                    ax.plot(xs, ys)
                elif style == 'tiling':
                    if j < len(row) - 1:
                        p2 = (row[j+1][0]*scale, row[j+1][1]*scale)
                        xs, ys = line_between_dots((x_s,y_s), p2)
                        ax.plot(xs, ys)
                    if i < len(dot_grid) - 1:
                        p2 = (dot_grid[i+1][j][0]*scale, dot_grid[i+1][j][1]*scale)
                        xs, ys = line_between_dots((x_s,y_s), p2)
                        ax.plot(xs, ys)
                elif style == 'cross':
                    segs = cross_lines((x_s, y_s), length=0.35*scale)
                    for xs, ys in segs:
                        ax.plot(xs, ys)
                idx += 1
        return artists

    ani = animation.FuncAnimation(fig, draw_frame, frames=max_frames, interval=interval, blit=False, repeat=False)
    return ani

File: renderer.js
// renderer.js - Canvas rendering and animation

class Renderer {
    constructor(canvas) {
        this.canvas = canvas;
        this.ctx = canvas.getContext('2d');
        this.animationId = null;
        this.isAnimating = false;
    }

    clear() {
        this.ctx.clearRect(0, 0, this.canvas.width, this.canvas.height);
    }

    renderGrid(grid, style = 'loop', animate = false) {
        if (this.animationId) {
            cancelAnimationFrame(this.animationId);
            this.isAnimating = false;
        }

        this.clear();
        
        if (!grid || grid.length === 0) {
            this.drawPlaceholder();
            return;
        }

        // Center and scale the grid to fit canvas
        const centeredGrid = GridUtils.centerGrid(grid, this.canvas.width, this.canvas.height);
        
        if (animate) {
            this.animateKolam(centeredGrid, style);
        } else {
            this.drawStaticKolam(centeredGrid, style);
        }
    }

    drawStaticKolam(grid, style) {
        this.setupDrawingStyle();
        
        // Draw dots first
        this.ctx.fillStyle = '#2d3748';
        grid.forEach(row => {
            row.forEach(([x, y]) => {
                Motifs.drawDot(this.ctx, [x, y], 3);
            });
        });

        // Draw motifs
        this.ctx.strokeStyle = '#4299e1';
        this.ctx.lineWidth = 2;
        
        grid.forEach((row, i) => {
            row.forEach(([x, y], j) => {
                this.drawMotif([x, y], style, grid, i, j);
            });
        });
    }

    animateKolam(grid, style) {
        this.isAnimating = true;
        let currentDot = 0;
        const totalDots = grid.reduce((sum, row) => sum + row.length, 0);
        const animationSpeed = 100; // milliseconds per dot

        const animate = () => {
            if (!this.isAnimating) return;

            this.clear();
            this.setupDrawingStyle();

            // Draw all dots
            this.ctx.fillStyle = '#2d3748';
            grid.forEach(row => {
                row.forEach(([x, y]) => {
                    Motifs.drawDot(this.ctx, [x, y], 3);
                });
            });

            // Draw motifs up to current dot
            this.ctx.strokeStyle = '#4299e1';
            this.ctx.lineWidth = 2;
            
            let dotIndex = 0;
            for (let i = 0; i < grid.length; i++) {
                for (let j = 0; j < grid[i].length; j++) {
                    if (dotIndex >= currentDot) break;
                    
                    const [x, y] = grid[i][j];
                    this.drawMotif([x, y], style, grid, i, j);
                    dotIndex++;
                }
                if (dotIndex >= currentDot) break;
            }

            currentDot++;
            
            if (currentDot <= totalDots) {
                setTimeout(() => {
                    this.animationId = requestAnimationFrame(animate);
                }, animationSpeed);
            } else {
                this.isAnimating = false;
            }
        };

        animate();
    }

    drawMotif(center, style, grid, i, j) {
        switch (style) {
            case 'loop':
                Motifs.loopAroundDot(this.ctx, center, 25);
                break;
            case 'cross':
                Motifs.crossLines(this.ctx, center, 20);
                break;
            case 'diagonal':
                Motifs.diagonalCross(this.ctx, center, 20);
                break;
            case 'arc':
                const direction = (i + j) % 2 === 0 ? 'right' : 'left';
                Motifs.arcMotif(this.ctx, center, 20, direction);
                break;
            case 'spiral':
                Motifs.spiral(this.ctx, center, 2.5, 20);
                break;
            case 'tiling':
                // Connect to right neighbor
                if (j < grid[i].length - 1) {
                    const rightNeighbor = grid[i][j + 1];
                    Motifs.lineBetweenDots(this.ctx, center, rightNeighbor);
                }
                // Connect to bottom neighbor
                if (i < grid.length - 1 && j < grid[i + 1].length) {
                    const bottomNeighbor = grid[i + 1][j];
                    Motifs.lineBetweenDots(this.ctx, center, bottomNeighbor);
                }
                break;
        }
    }

    setupDrawingStyle() {
        this.ctx.lineCap = 'round';
        this.ctx.lineJoin = 'round';
        this.ctx.shadowColor = 'rgba(66, 153, 225, 0.3)';
        this.ctx.shadowBlur = 3;
        this.ctx.shadowOffsetX = 1;
        this.ctx.shadowOffsetY = 1;
    }

    drawPlaceholder() {
        this.ctx.fillStyle = '#a0aec0';
        this.ctx.font = '24px Arial';
        this.ctx.textAlign = 'center';
        this.ctx.textBaseline = 'middle';
        this.ctx.fillText(
            'Generate a kolam or upload an image',
            this.canvas.width / 2,
            this.canvas.height / 2
        );
    }

    stopAnimation() {
        this.isAnimating = false;
        if (this.animationId) {
            cancelAnimationFrame(this.animationId);
            this.animationId = null;
        }
    }

    downloadCanvas(filename = 'kolam.png') {
        const link = document.createElement('a');
        link.download = filename;
        link.href = this.canvas.toDataURL();
        link.click();
    }
}